### DISCUSSION TITLE
"Please tell me about the error with FindVariableFeatures. ~Error in '.SelectFeatures（）

### DISCUSSION THREAD

**Discussion by keimiyako** [satijalab/seurat#8233] [Q&A]
Hi all! I'm very new to Seurat and scRNA-seq data...
I'm getting a weird error that I have no idea how to fix.
I also couldn't find information about this error.

I've put in my data filtered_feature_bc_matrix (barcodes.tsv.gz, matrixmtx.gz, and features.tsv.gz) as a Seurat object beginning with this:

ctrl.data <- Read10X(data.dir =
"~/Desktop/scRNAseq/outs/filtered_feature_bc_matrix") 
ctrl <- CreateSeuratObject(counts = ctrl.data, project = “IMMUNE_CTRL”, min.cells = 5)
ctrl$stim <- “CTRL” 
ctrl$replicate <- “n=1”
ctrl <- subset(ctrl, subset = nFeature_RNA > 500)
ctrl <- NormalizeData(ctrl, verbose = FALSE)
ctrl <- FindVariableFeatures(ctrl, selection.method = "vst", nfeatures = 2000)

Yet Continue to get this error each time:

Finding variable features for layer counts.Gene Expression
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Finding variable features for layer counts.Multiplexing Capture
Calculating gene variances
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Calculating feature variances of standardized and clipped values
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
Error in `.SelectFeatures()`:
! None of the features provided are present in the feature set
Run `rlang::last_trace()` to see where the error occurred.
 15 件の警告がありました (警告を見るには warnings() を使って下さい) 
> rlang::last_trace()
<error/rlang_error>
Error in `.SelectFeatures()`:
! None of the features provided are present in the feature set
---
Backtrace:
    ▆
 1. ├─Seurat::FindVariableFeatures(...)
 2. └─Seurat:::FindVariableFeatures.Seurat(...)
 3.   ├─Seurat::FindVariableFeatures(...)
 4.   └─Seurat:::FindVariableFeatures.StdAssay(...)
 5.     ├─SeuratObject::VariableFeatures(...)
 6.     └─SeuratObject:::VariableFeatures.Assay5(...)
 7.       ├─SeuratObject::.SelectFeatures(...)
 8.       └─SeuratObject:::.SelectFeatures.list(...)
Run rlang::last_trace(drop = FALSE) to see 1 hidden frame.

I keep encountering the same error consistently. 
I'm having a tough time with this function and trying to understand how to fix this error.
How can I resolve this issue? 
I really appreciate your help!

**Comment by bioshot-dotcom** [comment]
I had the same error as you, for me it worked removing any data not related to gene expression after reading them with Read10X()...

**Reply by gupta567varun** [reply]
How did you remove the data not related to gene expression. I have within my RNA assay, 2 different count matrices, one for rna expression and one for antibody capture. I want to remove my antibody capture count from RNA assay. How can I do this?

**Reply by bioshot-dotcom** [reply]
You can either load data in this way `cts.expr<-Read10X("path/data")[["Gene Expression"]]` or set to `NULL` the antibody capture matrix in the object

**Comment by Ryan-Zhengyuan** [comment]
I also encountered  the same error, while my seurat object consisted of 6 data layers. In my case, I found two layers contained the same features (the two layers derived from the same dataset), so I removed the duplicated layers. It worked in the end by processing the object with the last 5 layers.

**Reply by Emily-K-Payne** [reply]
I am new to R and running into the same issue. How were you able to delete the duplicated?

**Reply by Ryan-Zhengyuan** [reply]
Previously, I had 6 datasets, and I created a Seurat object for each of them. After merging 6 objects, I obtained one object with 6 layers. So, what I did was just remove one of the 6 datasets and re-merge an object with 5 layers. Sorry that I cannot provide you with further suggestions because I'm also new to Seurat.

Good luck

**Comment by zqumich** [comment]
I had the same error as you. Have you solved the problem.

**Comment by aaltulea** [comment]
I faced this issue a lot when integrating multiple objects. This is how I fixed it. Essentially you need to remove the faulty layers. 
```
# obtain a list of variable features from another seurat object. Use an object that you know is of a good quality
var_feat = VariableFeatures(reference_seurat)
# find all features per layer
features_rna_per_layer = seurat@assays$RNA@features
# only keep features that are matching the reference features
features_rna_per_layer = features_rna_per_layer[rownames(features_rna_per_layer) %in% var_feat,]
# see how many layers have how many matches aganist the reference genes
true_counts_per_layer = colSums(features_rna_per_layer == TRUE)
# here you can note the layers that have unusual numbers, such as < 1400, since the reference geneset is 2000
hist(true_counts_per_layer, breaks = 100)
```
![image](https://github.com/satijalab/seurat/assets/100591786/1fee3c91-ef9f-4e26-a7fa-74c7eedb332b)

```
# a list of bad layers using a cutoff of 1400, since the reference geneset is 2000
bad_layers = true_counts_per_layer[true_counts_per_layer < 1400] %>% names()
# also make sure to remove data, scale.data, in addition to counts
bad_layers = stringr::str_replace_all(bad_layers, pattern = "^data.", "counts.")
bad_layers = c(bad_layers, stringr::str_replace_all(bad_layers, pattern = "^counts.", "data."))
bad_layers = c(bad_layers, stringr::str_replace_all(bad_layers, pattern = "^data.", "scale.data."))
# remove the layers from your main object
seurat@assays$RNA@layers[bad_layers] = NULL
# now FindVariableFeatures works without an issue
seurat = FindVariableFeatures(seurat)

```

**Comment by metaspine** [comment]
Maybe your seurat object did not contain genes when you create.
you can check by "rownames(seurat)",if the return is not gene name, you need to check you input(from the step of creat CreateSeuratObject())

**Reply by zchczc** [reply]
> Maybe your seurat object did not contain genes when you create. you can check by "rownames(seurat)",if the return is not gene name, you need to check you input(from the step of creat CreateSeuratObject())
Thank you! I found the reason for the error.

**Reply by Wanrenhua** [reply]
> > Maybe your seurat object did not contain genes when you create. you can check by "rownames(seurat)",if the return is not gene name, you need to check you input(from the step of creat CreateSeuratObject())
> > Thank you! I found the reason for the error.

Can you tell me how to resolve this error?Cause I meet this error,too.

**Comment by rersister** [comment]
I have the same questions.  I solved it from remove  layers. First you need to check  your  obj layers.  I have this question because I merged tow datasets. The combined obj show followed.  
![微信图片_20250611170817](https://github.com/user-attachments/assets/630e03de-8cef-4c4d-956c-7da5c24233a6)
The red box layers are wrong. Then remove them. What I do is the following code.

# combined is the Seurat object
# get layer name
layer_names <- names(combined@assays[["RNA"]]@layers)
# remove the wrong layer 
remove_layers <- layer_names[4:6]
# remove them 
for (layer in remove_layers) {
  combined@assays[["RNA"]]@layers[[layer]] <- NULL
}
#check layers
names(combined@assays[["RNA"]]@layers)

Just by do this,  rerun  NormalizeData and FindVariableFeatures function. All will OK.
